### PR TITLE
Adding the resizable content hover code without the persistance mechanism

### LINKED ISSUE #14165: Configure tooltip max width
When you hover over a type in a TypeScript file, the tooltip that appears has a very low max width, making it difficult to read long type signatures.

### PR DESCRIPTION
**Pull request by aiday-mar**
Fixes https://github.com/microsoft/vscode/issues/14165

This PR builds atop the initial PR on making content hovers resizable. The aim of this PR is to separate the code which does the persistance from the code which does the resizing. The persistance mechanism can be added later.

### CODE REVIEW

File: src/vs/editor/contrib/hover/browser/hover.ts [hunk 1]
```diff
@@ -157,15 +157,15 @@ export class ModesHoverController implements IEditorContribution {
...
+			!this._isHoverSticky && target.type === MouseTargetType.CONTENT_WIDGET && target.detail === ResizableHoverWidget.ID
 			&& this._contentWidget?.isColorPickerVisible()
 		) {
 			// though the hover is not sticky, the color picker needs to.
 			return;
 		}
 
-		if (this._isHoverSticky && target.type === MouseTargetType.OVERLAY_WIDGET && target.detail === MarginHoverWidget.ID) {
-			// mouse moved on top of overlay hover widget
+		if (target.type === MouseTargetType.OVERLAY_WIDGET && target.detail === MarginHoverWidget.ID) {
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/hover.ts:167]: Is the removal of `this._isHoverSticky` intentional?

File: src/vs/editor/contrib/hover/browser/hover.ts [hunk 2]
```diff
@@ -222,7 +222,9 @@ export class ModesHoverController implements IEditorContribution {
 
 		this._hoverClicked = false;
 		this._glyphWidget?.hide();
-		this._contentWidget?.hide();
+		if (!this._contentWidget?.widget.isResizing) {
+			this._contentWidget?.hide();
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/hover.ts:226]: Can we add a comment explaining why the widget doesn't hide while resizing. Or better explain why we being hidden while resizing?

File: src/vs/editor/contrib/hover/browser/contentHover.ts [hunk 3]
```diff
@@ -68,16 +67,22 @@ export class ContentHoverController extends Disposable {
...
+	get widget() {
+		return this._widget;
+	}
+
 	/**
 	 * Returns true if the hover shows now or will show.
 	 */
 	public maybeShowAt(mouseEvent: IEditorMouseEvent): boolean {
+		if (this._widget.isResizing) {
+			return true;
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/contentHover.ts:84]: Similar to hiding and resizing. Is it that during resize we get show/hide requests?

File: src/vs/editor/contrib/hover/browser/contentHover.ts [hunk 4]
```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
...
+		if (this._hasHorizontalScrollbar()) {
+			this._adjustContentsBottomPadding();
+			this._adjustHoverHeightForScrollbar(clientHeight);
+		}
+		this._layoutContentWidget();
 	}
 
 	public focus(): void {
-		this._hover.containerDomNode.focus();
+		this._hoverWidget.containerDomNode.focus();
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/contentHover.ts:781]: suggestion: there are a bunch of renames that make this PR larger than needed (`_hoverData`, `ResizableHoverWidget`, `_hoverWidget` etc.) No feelings about the changes per se but I would suggest you do them in a separate PR so that this PR stays focused to resizing and so that the GIT history is more precise

File: src/vs/editor/contrib/hover/browser/contentHover.ts [hunk 5]
```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
...
 		}
 	}
 
-	public clear(): void {
-		this._hover.contentsDomNode.textContent = '';
+	private _setPersistedHoverDimensionsOrRenderNormally(): void {
+		// Added because otherwise the initial size of the hover content is smaller than should be
+		const layoutInfo = this._editor.getLayoutInfo();
+		this._resizableNode.layout(layoutInfo.height, layoutInfo.width);
+		this._setHoverWidgetDimensions('auto', 'auto');
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/contentHover.ts:748]: This confuses me a little because it sets the hover to be really large (the editor size) and then triggers an "auto-layout" to its actual size?

File: src/vs/editor/contrib/hover/browser/resizableContentWidget.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		return this._contentPosition;
+	}
+
+	get position(): Position | undefined {
+		return this._position;
+	}
+
+	set position(position: Position | undefined) {
+		this._position = position;
+	}
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/resizableContentWidget.ts:68]: suggestion move this up into the ContentWidget because it seems specific to the content hover and not to resizable hovers in general

File: src/vs/editor/contrib/hover/browser/resizableContentWidget.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		this._resizableNode.layout(dimension.height, dimension.width);
+	}
+
+	beforeOnDidWillResize() {
+		return;
+	}
+
+	afterOnDidResize() {
+		return;
+	}
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/resizableContentWidget.ts:121]: these never seem to be used (no caller, no implementor)

File: src/vs/editor/contrib/hover/browser/contentHover.ts [hunk 8]
```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
...
-		if (hasHorizontalScrollbar) {
-			// There is just a horizontal scrollbar
-			const extraBottomPadding = `${this._hover.scrollbar.options.horizontalScrollbarSize}px`;
-			if (this._hover.contentsDomNode.style.paddingBottom !== extraBottomPadding) {
-				this._hover.contentsDomNode.style.paddingBottom = extraBottomPadding;
-				this._editor.layoutContentWidget(this);
-				this._hover.onContentsChanged();
-			}
+		if (this._hoverData.stoleFocus) {
+			this._editor.focus();
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/contentHover.ts:740]: While testing your changes I saw this erroring (see below). That's because of set `_setHoverData` being called a few lines above

```ERR Cannot read properties of undefined (reading 'stoleFocus'): TypeError: Cannot read properties of undefined (reading 'stoleFocus')
    at ResizableHoverWidget.hide (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:602:33)
    at ContentHoverController._setCurrentResult (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:161:30)
    at ContentHoverController.hide (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/contentHover.js:167:18)
    at ModesHoverController._hideWidgets (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:162:38)
    at ModesHoverController._onEditorMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:94:18)
    at vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/contrib/hover/browser/hover.js:56:70
    at Listener.invoke (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:660:27)
    at EventDeliveryQueue.deliver (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:836:38)
    at InteractionEmitter.fire (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/base/common/event.js:800:37)
    at InteractionEmitter.fire (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/widget/codeEditorWidget.js:1431:19)
    at viewUserInputEvents.onMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/widget/codeEditorWidget.js:1318:74)
    at ViewUserInputEvents.emitMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/view/viewUserInputEvents.js:37:32)
    at ViewController.emitMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/view/viewController.js:248:34)
    at TouchHandler._onMouseLeave (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/controller/mouseHandler.js:193:33)
    at vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/controller/mouseHandler.js:40:94
    at HTMLDivElement.<anonymous> (vscode-file://vscode-app/Users/jrieken/Code/_collab/vscode2/out/vs/editor/browser/editorDom.js:131:17)
```

File: src/vs/editor/contrib/hover/browser/contentHover.ts [hunk 9]
```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
...
-	public readonly allowEditorOverflow = true;
+	public static ID = 'editor.contrib.resizableContentHoverWidget';
 
-	private readonly _hoverVisibleKey = EditorContextKeys.hoverVisible.bindTo(this._contextKeyService);
-	private readonly _hoverFocusedKey = EditorContextKeys.hoverFocused.bindTo(this._contextKeyService);
-	private readonly _hover: HoverWidget = this._register(new HoverWidget());
-	private readonly _focusTracker = this._register(dom.trackFocus(this.getDomNode()));
-	private readonly _horizontalScrollingBy: number = 30;
-	private _visibleData: ContentHoverVisibleData | null = null;
+	private _disposableStore = new DisposableStore();
```
Comment (jrieken) [inline src/vs/editor/contrib/hover/browser/contentHover.ts:453]: Bug: This isn't actually disposed. Also you inhert a store (and _register-function) from the super class that could be using

File: src/vs/editor/contrib/hover/browser/hover.ts (hunk 2 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/hover/browser/hover.ts]: Perhaps it is better to not call _hideWidgets at all when we are resizing, instead I am doing the isResizing check in the various places where _hideWidgets is called, in the onMouseLeave and onMouseMove events

File: src/vs/editor/contrib/hover/browser/contentHover.ts (hunk 3 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/hover/browser/contentHover.ts:84]: This method is called inside of _onMouseMove to decide whether the mouse is showing or will show. In the case when the widget is resizing, the hover is already showing.

File: src/vs/editor/contrib/hover/browser/contentHover.ts (hunk 5 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/hover/browser/contentHover.ts:748]: I don't recall why I needed this code, I think it was to not make the hover widget size be limited by the resizable node size. I will try without this code.

File: src/vs/editor/contrib/hover/browser/hover.ts (hunk 1 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/hover/browser/hover.ts]: I think this was unintentional

File: src/vs/editor/contrib/hover/browser/resizableContentWidget.ts (hunk 7 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/hover/browser/resizableContentWidget.ts:121]: That's right, this is code from the initial PR which persists the hover sizes.

**Review by jrieken — DISMISSED** [review]

**Review by jrieken — APPROVED** [review]
